src/java.desktop/macosx/classes/sun/awt/CGraphicsDevice.java

```diff
@@ -65,9 +66,31 @@
 
     public CGraphicsDevice(final int displayID) {
         this.displayID = displayID;
-        config = MacOSFlags.isMetalEnabled() ?
-                MTLGraphicsConfig.getConfig(this, displayID, 0) :
-                CGLGraphicsConfig.getConfig(this);
+
+        // Check whether -Dsun.java2d.metal=true has been specified
+        if (MacOSFlags.isMetalEnabled()) {
```

Comment:
Is it possible to use metal and OGL at the same time on different devices? I think it is better to move these checks to the CGraphicsEnvironment.makeScreenDevice(), similar Win32GraphicsEnvironment.makeScreenDevice()

Reply:
I am not sure why someone wants to use different pipelines on different devices? and should we support it?
On Windows we do that as a fallback to overcome a limitation of WGL - that's what I could infer from code in W32GraphicsDevice.getDefaultConfiguration()

I doubt whether we can add these checks in CGraphicsEnvironment.makeScreenDevice() - as we do not have specific MTLGraphicsDevice and CGLGraphicsDevice classes. That's the reason this method is currently unsupported.

Win32GraphicsEnvironment.makeScreenDevice() - creates either D3DGraphicsDevice or Win32GraphicsDevice based on whether isD3DEnabled. 
Win32GraphicsDevice in turn uses either WGLGraphicsConfig or Win32GraphicsConfig based on whether OGL is enabled.

I felt CGraphicsDevice (similar to Win32GraphicsDevice) can differentiate between metal or OGL. Hence, I have added checks here.

Reply:
If it is unsupported then why you should check that for every device creation? What happens if one device was created using metal and then another device will use OGL?

Reply:
I have moved the Metal framework availability check to be done only once now.
There is still a check if MTLGraphicsConfig.getConfig() fails by any chance. 

CGraphicsDevice contains GraphicsConfiguration - which is created in the constructor - It can be either MTLGraphicsConfiguration or OpenGLConfiguration. This check needs to be done while creating the config.
For Win32GraphicsDevice - the config is not created in the constructor, but at a later stage in getDefaultConfiguration(). So only the place of checking for OGL or falling back to GDI is different, but the concept is the same.

The whole idea of one device using metal and another device using OGL is hypothetical. The condition check cannot pass for one device and fail for another.

Reply:
I agree that it makes no sense to even consider using different pipelines on different devices. The change you made to do the check once seems best.